swap/src/network/tor_transport.rs

```diff
@@ -0,0 +1,225 @@
+            Ok(stream)
+        }
+
+        match to_address_string(addr.clone()) {
+            Some(tor_address_string) => {
+                Ok(Box::pin(do_tor_dial(self.socks_port, tor_address_string)))
+            }
+            _ => {
+                tracing::warn!(
+                    "Address {} could not be formatted. Dialling via clear net",
+                    addr
+                );
+                self.inner.dial(addr)
+            }
+        }
+    }
+
+    fn address_translation(&self, listen: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        self.inner.address_translation(listen, observed)
+    }
+}
+
+/// Tor expects an address format of ADDR:PORT.
+/// This helper function tries to convert the provided multi-address into this
+/// format. None is returned if an unsupported protocol was provided.
+fn to_address_string(multi: Multiaddr) -> Option<String> {
+    let components = multi.iter();
+    for protocol in components {
+        if let Protocol::Onion3(addr) = protocol {
+            return Some(format!(
+                "{}.onion:{}",
+                BASE32.encode(addr.hash()).to_lowercase(),
+                addr.port()
+            ));
+        }
+    }
+
+    // Deal with non-onion addresses
+    let protocols = multi.iter().collect::<Vec<_>>();
+    let address_string = protocols
+        .iter()
+        .filter_map(|protocol| match protocol {
+            Protocol::Ip4(addr) => Some(format!("{}", addr)),
+            Protocol::Ip6(addr) => Some(format!("{}", addr)),
+            Protocol::Dns(addr) => Some(format!("{}", addr)),
+            Protocol::Dns4(addr) => Some(format!("{}", addr)),
+            _ => None,
+        })
+        .collect::<Vec<_>>();
+    if address_string.is_empty() {
+        tracing::warn!(
+            "Could not format address {}. Please consider reporting this issue. ",
+            multi
+        );
+        return None;
+    }
+
+    let address_string = address_string
+        .get(0)
+        .expect("Valid multiaddr consist out of max 1 address")
+        .clone();
+
+    // check for port
+    let port = protocols
+        .iter()
+        .filter_map(|protocol| match protocol {
+            Protocol::Tcp(port) => Some(format!("{}", port)),
+            Protocol::Udp(port) => Some(format!("{}", port)),
+            _ => None,
+        })
+        .collect::<Vec<_>>();
+
+    match port.len().cmp(&1) {
+        Ordering::Greater => {
+            tracing::warn!(
+                "Did not expect more than 1 port in address {}. Please consider reporting this issue.",
+                multi
+            );
+            Some(address_string)
+        }
+        Ordering::Less => Some(address_string),
+        Ordering::Equal => Some(format!(
+            "{}:{}",
+            address_string,
+            port.get(0)
+                .expect("Already verified the length of the vec.")
+        )),
+    }
```

Comment:
That looks quite complicated. Any reason you can't just step through the iterator as needed?

```rust
let mut protocols = addr.iter();

let address = match protocols.next() {
	// extract ip4 etc, fail if non-matching
	// do proper error handling on `None`
};

let port = match protocols.next() {
	// extract port for tcp / udp etc, fail if non-matching
	// do proper error handling on `None`
};

format!("{}:{}", address, port)

```

Reply:
Also, a `Result` seems more appropriate if we _fail_ to extract a usable address.

Reply:
Agree, it was also cumbersome to write. :) 

Unfortunately I do not know how many protocols in the multi address might or might not be. Hence the iterator.

Reply:
Atm we do not fail but fall back to the default Tcp dialling.

Reply:
Check out the last commit #00b40c4a8fd92276d89517fcac4ee311f68e4b3e :)